src/Turtle/Prelude.hs

```diff
@@ -1986,3 +1991,45 @@ single s = do
         _   -> do
             let msg = format ("single: expected 1 line of input but there were "%d%" lines of input") (length as)
             die msg
+
+-- | Filter adjacent duplicate elements:
+--
+-- >>> view (uniq (select [1,1,2,1,3]))
+-- 1
+-- 2
+-- 1
+-- 3
+uniq :: Eq a => Shell a -> Shell a
+uniq s = Shell $ \(FoldM step begin done) -> do
+  ref <- newIORef Nothing
```

Comment:
I believe you can do this without an `IORef`.  I haven't type-checked this, but I believe it would go something like this:

```haskell
uniq s = Shell $ \(FoldM step begin done) -> do
    let step' (x, Just a') a | a == a' = do
            return (x, Just a)
        step' (x, _      ) a | otherwise = do
            x' <- step x a
            return (x', Just a)

    let begin' = (begin, Nothing)

    let done' (x, _) = done x

    foldIO s (FoldM step' begin' done')
```

This takes advantage of the fact that the existentially quantified state of the new `FoldM` doesn't have to have the same type as the existentially quantified state of the fold `FoldM`

Similar comment for `nub`

Reply:
Yeah that is *so* much better, or at least I can sleep better